libraries/vircadia-client/src/internal/Context.cpp

```diff
@@ -0,0 +1,165 @@
+            nodeListParams, userAgent, info
+        ] () {
+            setupHifiApplication(info.name, info.organization, info.domain, info.version);
+            Setting::init();
+
+            QCoreApplication qtApp{argc, &argv};
+
+            DependencyManager::registerInheritance<LimitedNodeList, NodeList>();
+
+            if (userAgent != nullptr) {
+                DependencyManager::set<AccountManager>(false,
+                    [userAgentString = QString(userAgent)]
+                    () { return userAgentString; });
+            } else {
+                DependencyManager::set<AccountManager>(false);
+            }
+            DependencyManager::set<DomainAccountManager>();
+            DependencyManager::set<AddressManager>();
+            DependencyManager::set<NodeList>(nodeListParams);
+
+            {
+                auto accountManager = DependencyManager::get<AccountManager>();
+                accountManager->setIsAgent(true);
+                accountManager->setAuthURL(MetaverseAPI::getCurrentMetaverseServerURL());
+            }
+
+            {
+                auto nodeList = DependencyManager::get<NodeList>();
+
+                // setup a timer for domain-server check ins
+                QTimer* domainCheckInTimer = new QTimer(nodeList.data());
+                QObject::connect(domainCheckInTimer, &QTimer::timeout, nodeList.data(), &NodeList::sendDomainServerCheckIn);
+                domainCheckInTimer->start(DOMAIN_SERVER_CHECK_IN_MSECS);
+
+                // start the nodeThread so its event loop is running
+                // (must happen after the checkin timer is created with the nodelist as it's parent)
+                nodeList->startThread();
+
+                // TODO: parameterize
+                nodeList->addSetOfNodeTypesToNodeInterestSet(NodeSet() << NodeType::AudioMixer << NodeType::AvatarMixer
+                     << NodeType::EntityServer << NodeType::AssetServer << NodeType::MessagesMixer);
+            }
+
+            // TODO: account manager login
+
+            QObject::connect(&qtApp, &QCoreApplication::aboutToQuit, [](){
+                DependencyManager::prepareToExit();
+
+                {
+                    auto nodeList = DependencyManager::get<NodeList>();
+
+                    // send the domain a disconnect packet, force stoppage of domain-server check-ins
+                    nodeList->getDomainHandler().disconnect("Quit");
+                    nodeList->setIsShuttingDown(true);
+
+                    // tell the packet receiver we're shutting down, so it can drop packets
+                    nodeList->getPacketReceiver().setShouldDropPackets(true);
+                }
+
+                QThreadPool::globalInstance()->clear();
+                QThreadPool::globalInstance()->waitForDone();
+
+                // TODO: these are still being used after this point, so need a better way to make sure all threads and pending operations have finished before destroying
```

Comment:
This TODO seems quite important. Is it included in a "to do" list (e.g., gitHub issue) somewhere?

Reply:
I though it's not strictly necessary to destroy these explicitly, since they should be taken care of when the DependencyManager is destroyed. Still important to better understand how the threading works in general and be able to wait for all the threads to finish somehow. I'll be looking into it more next week, to try to figure out why Unity Editor freezes when running the example project. Not using the issue tracker much, cause I'm still figuring things out and am the only one working on it for now, but hopefully I'll be able bring it all together and define some kind of a roadmap to go forward next week, possible in a form GitHub issues.